util/network/src/client_version.rs

```diff
@@ -0,0 +1,473 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+// Parity client string prefix
+const LEGACY_CLIENT_ID_PREFIX: &str = "Parity";
+const PARITY_CLIENT_ID_PREFIX: &str = "Parity-Ethereum";
+
+// Parity versions starting from this will accept block bodies requests
+// of 256 bodies
+const PARITY_CLIENT_LARGE_REQUESTS_VERSION: &str = "2.3.0";
+
+// Parity versions starting from this will accept service-transactions
+const SERVICE_TRANSACTIONS_VERSION: &str = "1.6.0";
+
+use semver::Version;
+use std::fmt;
+
+
+/// Description of the software version running in a peer
+/// according to https://github.com/ethereum/wiki/wiki/Client-Version-Strings
+/// This structure as it is represents the format used by Parity clients. Other
+/// vendors may provide additional fields.
+///
+/// TODO support formats with extra fields, e.g.:
+/// "Geth/main.jnode.network/v1.8.21-stable-9dc5d1a9/linux-amd64/go1.11.4"
+
+#[derive(Clone,Debug,PartialEq,Eq)]
+pub struct ParityClientData {
+	name: String,
+	variant: Option<String>,
+	semver: Version,
+	os: String,
+	compiler: String,
+}
+
+// Accessor methods for ParityClientData. This will probably
+// need to be abstracted away into a trait.
+impl ParityClientData {
+	fn name(&self) -> &str {
+		self.name.as_str()
+	}
+
+	fn variant(&self) -> Option<&str> {
+		self.variant.as_ref().map(String::as_str)
+	}
+
+	fn semver(&self) -> &Version {
+		&self.semver
+	}
+
+	fn os(&self) -> &str {
+		self.os.as_str()
+	}
+
+	fn compiler(&self) -> &str {
+		self.compiler.as_str()
+	}
+}
+
+#[derive(Clone,Debug,PartialEq,Eq)]
+pub enum ClientVersion {
+	ParityClient(
+		ParityClientData
+	),
+	ParityUnknownFormat(String),
+	Other(String), // Id string
+}
+
+// TODO: Maybe merge with Peercapabilityinfo in ethcore-network?
+pub trait ClientCapabilities {
+	fn can_handle_large_requests(&self) -> bool;
+
+	fn accepts_service_transaction(&self) -> bool;
+}
+
+// This is an implementation of a function taken from propagator.rs
+fn parity_accepts_service_transaction(parity_client_data: &ParityClientData) -> bool {
+	let service_transactions_version = Version::parse(SERVICE_TRANSACTIONS_VERSION).unwrap();
+
+	*parity_client_data.semver() >= service_transactions_version
+}
+
+impl ClientCapabilities for ClientVersion {
+	fn can_handle_large_requests(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(data) => {
+				if *data.semver() < Version::parse(PARITY_CLIENT_LARGE_REQUESTS_VERSION).unwrap() {
```

Comment:
I considered the option of making ClientVersion a struct and have a field for the provider instead of variants Parity/Geth/Other. When making choices about "what kind of ..." I prefer pattern matching instead of ifs.

Sure, in this case I would try to avoid parsing and comparing the whole time too. Since the comparison is only made in case it's parity, I would add a constructor and a field to ParityClientData for now, and leave ClientVersion as an enum. What do you think?

Reply:
To be clear, I was actually thinking about preserving the enum and creating a wrapping struct, that contains an enum and a field `can_handle_large_requests`, which is initialized in the constructor.

Reply:
sounds good to me